Compiling ARM64 error after installing Xcode15

### Godot version

4.2.dev [e3e2528ba]

### System information

Godot v4.2.dev (e3e2528ba) - macOS 14.0.0 - Vulkan (Forward+) - integrated Apple M1 - Apple M1 (8 Threads)

### Issue description

When I wanted to compile Godot for Apple Silicon today, I have encountered such an error:

```shell
scons: Reading SConscript files ...
Auto-detected 8 CPU cores available for build parallelism. Using 7 cores by default. You can override it with the -j argument.
Building for macOS 11.0+.
MoltenVK found at: ../MoltenVK/MoltenVK/MoltenVK.xcframework/macos-arm64_x86_64/
Building for platform "macos", architecture "arm64", target "editor".
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[ 96%] Linking Program bin/godot.macos.editor.arm64 ...
[100%] progress_finish(["progress_finish"], [])
[100%] 0  0x10069f648  __assert_rtn + 72
1  0x1005c7c5c  ld::Fixup::applyFixup(ld::Atom const*, ld::LayoutLinkedImage const&, unsigned char*) const + 8268
2  0x10065a7d8  ___ZN2ld16LayoutExecutable27writeContentWithoutLinkEditENSt3__14spanIhLm18446744073709551615EEEy_block_invoke + 332
3  0x18ed51950  _dispatch_client_callout2 + 20
4  0x18ed64ba0  _dispatch_apply_invoke + 176
5  0x18ed51910  _dispatch_client_callout + 20
6  0x18ed633cc  _dispatch_root_queue_drain + 864
7  0x18ed63a04  _dispatch_worker_thread2 + 156
8  0x18eefb0d8  _pthread_wqthread + 228
ld: Assertion failed: (extras.otherInstrOffset != 0 && "Kind::arm64_adrp_ldr missing extra info"), function applyFixup, file Fixup.cpp, line 793.
clang: error: linker command failed with exit code 1 (use -v to see invocation)
scons: *** [bin/godot.macos.editor.arm64] Error 1
scons: building terminated because of errors.
```


### Steps to reproduce

It's the command I used: 

```shell
scons platform=macos arch=arm64 target=editor vulkan_sdk_path='../MoltenVK'
```

I'm sure there was no problem until I updated XCode15, but `x86_64` is done:

```shell
scons platform=macos arch=x86_64 target=editor vulkan_sdk_path='../MoltenVK'

scons: Reading SConscript files ...
Auto-detected 8 CPU cores available for build parallelism. Using 7 cores by default. You can override it with the -j argument.
Building for macOS 10.13+.
MoltenVK found at: ../MoltenVK/MoltenVK/MoltenVK.xcframework/macos-arm64_x86_64/
Building for platform "macos", architecture "x86_64", target "editor".
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[100%] progress_finish(["progress_finish"], [])
[100%] scons: done building targets.
```

### Minimal reproduction project

So what should I do to build the arch of arm64?

I am also having a similar problem. But in my case, for some reason, it's not finding the required Vulkan symbols to compile to arm64. I am using Xcode 12.5
```bash
scons -j4 target=editor/template_debug/template_release arch=arm64
Automatically detected platform: macos
Building for macOS 11.0+.
MoltenVK found at: /Users/macosgodot/VulkanSDK/1.3.261.1/MoltenVK/MoltenVK.xcframework/macos-arm64_x86_64/
Building for platform "macos", architecture "arm64", target "template_debug".
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[ 97%] Linking Program bin/godot.macos.template_debug.arm64 ...
[100%] progress_finish(["progress_finish"], [])
[100%] Undefined symbols for architecture arm64:
  "_objc_msgSend$isReadableFileAtPath:", referenced from:
      mvk::canReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$stringByAppendingString:", referenced from:
      mvk::pathWithExtension(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$stringByAppendingPathExtension:", referenced from:
      mvk::pathWithExtension(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$lastPathComponent", referenced from:
      mvk::fileName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$stringByDeletingPathExtension", referenced from:
      mvk::fileName(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
      mvk::pathWithExtension(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$isWritableFileAtPath:", referenced from:
      mvk::canWriteFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
  "_objc_msgSend$isAbsolutePath", referenced from:
      mvk::absolutePath(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libMoltenVK.a(libMoltenVK.a-arm64-master.o)
```
I already tried re-installing the Vulkan SDK and that did not work.

--- a/platform/macos/detect.py
+++ b/platform/macos/detect.py
@@ -1,6 +1,6 @@
 import os
 import sys
-from methods import detect_darwin_sdk_path
+from methods import detect_darwin_sdk_path, get_compiler_version, is_vanilla_clang
 from platform_methods import detect_arch
 
 from typing import TYPE_CHECKING
@@ -118,6 +118,22 @@ def configure(env: "Environment"):
         env.Append(ASFLAGS=["-arch", "x86_64", "-mmacosx-version-min=10.13"])
         env.Append(CCFLAGS=["-arch", "x86_64", "-mmacosx-version-min=10.13"])
         env.Append(LINKFLAGS=["-arch", "x86_64", "-mmacosx-version-min=10.13"])
+
+    cc_version = get_compiler_version(env) or {
+        "major": None,
+        "minor": None,
+        "patch": None,
+        "metadata1": None,
+        "metadata2": None,
+        "date": None,
+    }
+    cc_version_major = int(cc_version["major"] or -1)
+    cc_version_minor = int(cc_version["minor"] or -1)
+    vanilla = is_vanilla_clang(env)
+
+    # Workaround for Xcode 15 linker bug.
+    if not vanilla and cc_version_major == 15 and cc_version_minor == 0:
+        env.Prepend(LINKFLAGS=["-ld_classic"])
 
     env.Append(CCFLAGS=["-fobjc-arc"])
 

